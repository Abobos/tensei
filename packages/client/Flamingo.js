import Axios from 'axios'
import { Notification } from '@contentful/forma-36-react-components'

import TextField from './fields/Text'
import DateField from './fields/Date'
import HasManyField from './fields/HasMany'
import LinkField from './index-fields/Link'
import DateIndexField from './index-fields/Date'

import TextIndexField from './index-fields/Text'

class Flamingo {
    state = (() => {
        let user = null
        let resources = []
        let appConfig = {}
        let shouldShowRegistrationScreen = false

        try {
            appConfig = JSON.parse(window.flamingoDefaultState.appConfig)
            resources = JSON.parse(window.flamingoDefaultState.resources)
            shouldShowRegistrationScreen = window.flamingoDefaultState.shouldShowRegistrationScreen === "true"

            user = JSON.parse(window.flamingoDefaultState.user || null)

        } catch (errors) {}

        return {
            user,
            resources,
            appConfig,
            shouldShowRegistrationScreen
        }
    })()

    library = {
        Notification,
    }

    getPath = (path) => {
        return `/${this.state.appConfig.dashboardPath}/${path}`
    }

    request = Axios.create({
        baseURL: '/api/',
    })

    fieldComponents = {
        DateField,
        TextField,
        HasManyField,
        LinkField: TextField,
        NumberField: TextField,
    }

    indexFieldComponents = {
        LinkField,
        TextField: TextIndexField,
        IDField: TextIndexField,
        NumberField: TextField,
        DateField: DateIndexField,
    }
}

window.Flamingo = new Flamingo()
