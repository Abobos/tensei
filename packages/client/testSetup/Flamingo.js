import TextField from '~/fields/Text'
import SelectField from '~/fields/Select'
import HasManyField from '~/fields/HasMany'
import DateTimeField from '~/fields/DateTime'
import TextareaField from '~/fields/Textarea'
import BelongsToField from '~/fields/BelongsTo'

export default {
    getPath: jest.fn(() => 'string'),
    fieldComponents: {
        TextField,
        SelectField,
        HasManyField,
        DateTimeField,
        TextareaField,
        BelongsToField,
        LinkField: TextField,
        NumberField: TextField,
        IntegerField: TextField,
        DateField: DateTimeField,
    },
    library: {
        Notification: { success: jest.fn(), error: jest.fn() },
    },
}
