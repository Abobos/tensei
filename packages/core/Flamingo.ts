import Path from 'path'
import BodyParser from 'body-parser'
import ExpressSession from 'express-session'
import Express, { Application } from 'express'
import ClientController from './controllers/ClientController'
import AuthController from './controllers/auth/AuthController'

import {
    text,
    resource,
    Resource,
    FlamingoConfig,
    ResourceManager,
    SupportedDatabases,
    DatabaseRepositoryInterface,
} from '@flamingo/common'

class Flamingo {
    public app: Application = Express()
    public databaseClient: any = null
    private toolsBooted: boolean = false
    private databaseBooted: boolean = false
    private registeredApplication: boolean = false
    private databaseRepository: DatabaseRepositoryInterface|null = null

    private config: FlamingoConfig = {
        resources: [],
        adminTable: 'administrators',
        dashboardPath: 'flamingo',
        apiPath: 'api',
        scripts: [
            {
                name: 'flamingo.js',
                path: Path.resolve(__dirname, 'client', 'index.js'),
            },
        ],
        styles: [
            {
                name: 'flamingo.css',
                path: Path.resolve(__dirname, 'client', 'index.css'),
            },
        ],
        env: {
            port: process.env.PORT || 1377,
            database: (process.env.DATABASE as SupportedDatabases) || 'sqlite',
            sessionSecret: process.env.SESSION_SECRET || 'test-session-secret',
            databaseUrl:
                process.env.DATABASE_URL || 'mysql://root@127.0.0.1/flmg',
        },
    }

    public async register() {
        if (this.registeredApplication) {
            return this
        }

        await this.registerDatabase()

        // Please do not change this order. Super important so bugs are not introduced.
        this.registerMiddleware()
        this.registerAssetsRoutes()
        this.registerCoreRoutes()

        this.registeredApplication = true

        return this
    }

    public start() {
        this.app.listen(this.config.env.port, () => {
            console.log(
                `App listening on port http://localhost:${this.config.env.port}`
            )
        })

        return this
    }

    public dashboardPath(dashboardPath: string) {
        this.setValue('dashboardPath', dashboardPath)

        return this
    }

    public async registerDatabase() {
        if (this.databaseBooted) {
            return this
        }

        // if database === mysql | sqlite | pg, we'll use the @flamingo/knex package, with either mysql, pg or sqlite3 package
        // We'll require('@flamingo/knex') and require('sqlite3') for example. If not found, we'll install.
        const { Repository } = require('@flamingo/knex')

        const repository: DatabaseRepositoryInterface = new Repository(
            this.config.env
        )

        const client = await repository.setup(this.config)

        this.databaseClient = client
        this.databaseRepository = repository

        this.app.use(
            (
                request: Express.Request,
                response: Express.Response,
                next: Express.NextFunction
            ) => {
                request.db = repository

                next()
            }
        )

        this.databaseBooted = true

        return this
    }

    public apiPath(apiPath: string) {
        this.setValue('apiPath', apiPath)

        return this
    }

    public registerMiddleware() {
        this.app.use(BodyParser.json())

        const Store = require('connect-session-knex')(ExpressSession)

        this.app.use(
            (
                request: Express.Request,
                response: Express.Response,
                next: Express.NextFunction
            ) => {
                request.resources = this.config.resources
                request.administratorResource = this.administratorResource()
                request.resourceManager = new ResourceManager(
                    this.config.resources,
                    this.databaseRepository!
                )

                next()
            }
        )

        this.app.use(
            ExpressSession({
                secret: this.config.env.sessionSecret,
                store: new Store({
                    knex: this.databaseClient,
                }),
                resave: false,
                saveUninitialized: false,
            })
        )
    }

    private getApiPath = (path: string) => {
        return `/${this.config.apiPath}/${path}`
    }

    public registerCoreRoutes() {
        // The administration dashboard
        this.app.get(
            `/${this.config.dashboardPath}(/*)?`,
            ClientController.index
        )

        this.app.post(this.getApiPath('login'), AuthController.login)
        this.app.post(this.getApiPath('register'), AuthController.register)

        this.app.post(this.getApiPath('logout'), AuthController.logout)
    }

    public registerAssetsRoutes() {
        this.app.use(
            (
                request: Express.Request,
                response: Express.Response,
                next: Express.NextFunction
            ) => {
                // Set the app config on express here. Should probably get its own function soon.
                request.appConfig = this.config
                request.scripts = this.config.scripts
                request.styles = this.config.styles

                next()
            }
        )

        this.config.scripts.concat(this.config.styles).forEach((asset) => {
            this.app.get(
                `/${asset.name}`,
                (request: Express.Request, response: Express.Response) =>
                    response.sendFile(asset.path)
            )
        })
    }

    private setValue(key: keyof FlamingoConfig, value: any) {
        this.config = {
            ...this.config,
            [key]: value,
        }

        return this
    }

    public resources(resources: Array<Resource>) {
        this.setValue('resources', [this.administratorResource(), ...resources])

        return this
    }

    private administratorResource() {
        return resource('Administrator')
            .hideFromNavigation()
            .fields([
                text('Name'),
                text('Email').unique().searchable(),
                text('Password'),
            ])
            .beforeCreate((payload) => {
                const Bcrypt = require('bcryptjs')

                return {
                    ...payload,
                    password: Bcrypt.hashSync(payload.password),
                }
            })
    }
}

export const flamingo = (config = {}) => {
    return new Flamingo()
}

export default Flamingo
