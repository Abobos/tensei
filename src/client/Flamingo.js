import Axios from 'axios'
import { Notification } from '@contentful/forma-36-react-components'

import TextField from './fields/Text'
import DateField from './fields/Date'
import HasManyField from './fields/HasMany'
import LinkField from './index-fields/Link'
import DateIndexField from './index-fields/Date'

import TextIndexField from './index-fields/Text'

class Flamingo {
    state = (() => {
        let user = null
        let resources = []

        try {
            user = JSON.parse(window.flamingoDefaultState.user)
            resources = JSON.parse(window.flamingoDefaultState.resources)
        } catch (errors) {}

        return {
            user,
            resources,
        }
    })()

    library = {
        Notification
    }

    request = Axios.create({
        baseURL: '/api/',
    })

    fieldComponents = {
        DateField,
        TextField,
        HasManyField,
        LinkField: TextField,
        NumberField: TextField,
    }

    indexFieldComponents = {
        TextField: TextIndexField,
        IDField: TextIndexField,
        NumberField: TextField,
        LinkField,
        DateField: DateIndexField,
    }
}

window.Flamingo = new Flamingo()
