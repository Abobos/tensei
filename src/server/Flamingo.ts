import Path from 'path'
import Consola from 'consola'
import Express from 'express'
import { FlamingoServiceProviderInterface, Asset } from './typings/interfaces'
import FlamingoServiceProvider from './providers/FlamingoServiceProvider'

class Flamingo {
    private $serviceProvider: FlamingoServiceProviderInterface

    public scripts: Asset[] = [
        {
            name: 'flamingo.js',
            path: Path.join(__dirname, '..', 'client', 'index.js'),
        },
    ]

    public styles: Asset[] = [
        {
            name: 'flamingo.css',
            path: Path.join(__dirname, '..', 'client', 'index.css'),
        },
    ]

    /**
     * A service provider is a class used to define all configurations
     * for flamingo. Routes, Cards, Resources, Authorization etc
     *
     * @param $serviceProvider FlamingoServiceProviderInterface
     */
    constructor(
        public $root: string,
        Provider: typeof FlamingoServiceProvider = FlamingoServiceProvider
    ) {
        this.$serviceProvider = new Provider($root)
    }

    public async start() {
        this.registerAssetsRoutes()

        await this.$serviceProvider.register()

        // this.$serviceProvider.launchServer((config) => {
        //     Consola.success('Server started on port', config.port)
        // })
    }

    public registerAssetsRoutes() {
        this.$serviceProvider.app.use(
            (
                request: Express.Request,
                response: Express.Response,
                next: Express.NextFunction
            ) => {
                request.scripts = this.scripts
                request.styles = this.styles

                next()
            }
        )

        this.scripts.concat(this.styles).forEach((asset) => {
            this.$serviceProvider.app.get(
                `/${asset.name}`,
                (request: Express.Request, response: Express.Response) =>
                    response.sendFile(asset.path)
            )
        })
    }

    public script(script: Asset) {
        this.scripts = [...this.scripts, script]

        return this
    }

    public style(style: Asset) {
        this.styles = [...this.styles, style]

        return this
    }
}

export default Flamingo
