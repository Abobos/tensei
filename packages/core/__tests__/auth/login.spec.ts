import Path from 'path'
import Knex from 'knex'
import Faker from 'faker'
import Bcrypt from 'bcryptjs'
import Supertest from 'supertest'
import { flamingo } from '../../Flamingo'

const setup = () => {
    process.env.DATABASE = 'mysql'
    process.env.DATABASE_URI = 'mysql://root@127.0.0.1/flamingotestdb'

    return flamingo().register()
}

test('validates login data and returns error messages with a 422', async () => {
    const { app, databaseClient } = await setup()

    const client = Supertest(app)

    const response = await client.post('/api/login').send({})

    expect(response.status).toBe(422)
    expect(response.body).toMatchSnapshot()

    databaseClient.destroy()
})

test('returns a 422 if user does not exist in database', async () => {
    const { app, databaseClient } = await setup()

    const client = Supertest(app)

    const response = await client.post('/api/login').send({
        email: 'hey@unknown-user.io',
        password: 'password',
        rememberMe: true,
    })

    expect(response.status).toBe(422)
    expect(response.body).toMatchSnapshot()

    databaseClient.destroy()
})

test('returns a 200 and success message when correct credentials are passed', async () => {
    const { app, databaseClient } = await setup()

    const knex: Knex = databaseClient

    const client = Supertest(app)

    await knex('administrators').truncate()

    const user = {
        password: 'password',
        name: Faker.name.findName(),
        email: Faker.internet.exampleEmail(),
    }

    await knex('administrators').insert({
        name: user.name,
        email: user.email,
        password: Bcrypt.hashSync(user.password),
    })

    const response = await client.post('/api/login').send({
        email: user.email,
        password: user.password,
        rememberMe: true,
    })

    expect(response.status).toBe(200)
    expect(response.body).toMatchSnapshot()

    knex.destroy()
})
