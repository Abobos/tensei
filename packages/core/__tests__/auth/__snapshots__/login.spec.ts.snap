// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns a 200 and success message when correct credentials are passed 1`] = `
Object {
  "message": "Login successful.",
}
`;

exports[`returns a 422 if user does not exist in database 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "email",
      "message": "These credentials do not match our records.",
    },
  ],
  "message": "Validation failed.",
}
`;

exports[`validates login data and returns error messages with a 422 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "email",
      "message": "The email is required.",
      "validation": "required",
    },
    Object {
      "field": "password",
      "message": "The password is required.",
      "validation": "required",
    },
  ],
  "message": "Validation failed.",
}
`;
